src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,417 @@
+                    compare = directionMap[key] == OrderByDirection.Descending ? binder.CreateBinaryExpression(BinaryOperatorKind.LessThan, property, constant, true) : binder.CreateBinaryExpression(BinaryOperatorKind.GreaterThan, property, constant, true);
+                }
+                else
+                {
+                    compare = binder.CreateBinaryExpression(BinaryOperatorKind.GreaterThan, property, constant, true);
+                }
+
+                if (firstProperty)
+                {
+                    lastEquality = binder.CreateBinaryExpression(BinaryOperatorKind.Equal, property, constant, true);
+                    where = compare;
+                    firstProperty = false;
+                }
+                else
+                {
+                    Expression condition = Expression.AndAlso(lastEquality, compare);
+                    where = where == null ? condition : Expression.OrElse(where, condition);
+                    lastEquality = Expression.AndAlso(lastEquality, binder.CreateBinaryExpression(BinaryOperatorKind.Equal, property, constant, true));
+                }
+            }
+
+            Expression whereLambda = Expression.Lambda(where, param);
+            return ExpressionHelpers.Where(query, whereLambda, query.ElementType);
+        }
+
+        /// <summary>
+        /// Generates a dictionary with property name and property values specified in the skiptoken value.
+        /// </summary>
+        /// <param name="value">The skiptoken string value.</param>
+        /// <param name="context">The <see cref="ODataQueryContext"/> which contains the <see cref="IEdmModel"/> and some type information</param>
+        /// <returns>Dictionary with property name and property value in the skiptoken value.</returns>
+        private static IDictionary<string, object> PopulatePropertyValuePairs(string value, ODataQueryContext context)
+        {
+            Contract.Assert(context != null);
+
+            IDictionary<string, object> propertyValuePairs = new Dictionary<string, object>();
+            IList<string> keyValuesPairs = ParseValue(value, CommaDelimiter);
+            foreach (string pair in keyValuesPairs)
+            {
+                string[] pieces = pair.Split(new char[] { propertyDelimiter }, 2);
+                if (pieces.Length > 1 && !String.IsNullOrWhiteSpace(pieces[0]))
+                {
+                    object propValue = null;
+                    IEdmStructuredType type = context.ElementType as IEdmStructuredType;
+                    Debug.Assert(type != null);
+                    IEdmTypeReference propertyType = null;
+                    IEdmProperty property = type.FindProperty(pieces[0]);
+                    if (property != null)
+                    {
+                        propertyType = property.Type;
+                    }
+                    propValue = ODataUriUtils.ConvertFromUriLiteral(pieces[1], ODataVersion.V401, context.Model, propertyType);
```

Comment:
>ConvertFromUriLiteral [](start = 46, length = 21)

how does ConvertFromUriLiteral handle null propertyType?

Reply:
It would still go ahead and parse it to a string by guessing the type. It will work fine with null propertyType for all except enum type. With enum type, it would return the string. 

---
In reply to: [293039050](https://github.com/OData/WebApi/pull/1724#discussion_r293039050) [](ancestors = 293039050)